### PR TITLE
Added MultiplexedCodec (and support for delimited collections)

### PR DESCRIPTION
**Pull request by ajaychandran**
Added support for codecs for delimited collections using a generalized MultiplexedCodec.

### CODE REVIEW

File: src/main/scala/scodec/codecs/package.scala [hunk 1]
```diff
@@ -984,6 +984,51 @@ package object codecs {
...
+   * remaining bits (input to `deMux` on next iteration) until a decoding error is encountered or no more bits remain.
+   * The final return value is a vector of all decoded element values.
+   *
+   * @param mux element multiplexer
+   * @param deMux element de-multiplexer (should return the next bits to decode and the remaining bits for next iteration)
+   * @param valueCodec element codec (used to decode next bits)
+   * @tparam A element type
+   * @return
+   */
+  def vectorMultiplexed[A](mux: (BitVector, BitVector) => BitVector, deMux: BitVector => (BitVector, BitVector), valueCodec: Codec[A]): Codec[Vector[A]] =
```
Comment (mpilquist) [inline src/main/scala/scodec/codecs/package.scala]: Needs `@group combinators` tag

File: src/main/scala/scodec/codecs/package.scala [hunk 2]
```diff
@@ -984,6 +984,51 @@ package object codecs {
...
+   * Example:
+   * {{{
+   * val codec = vectorDelimited(BitVector(' '), ascii)
+   * codec.decode(ascii.encode("i am delimited").require).require.value // Vector("i", "am", "delimited")
+   * }}}
+   *
+   * @param delimiter the bits used to separate element bit values
+   * @param valueCodec element codec (used to decode next bits)
+   * @tparam A element type
+   * @return
```
Comment (mpilquist) [inline src/main/scala/scodec/codecs/package.scala]: Needs `@group combinators` tag

File: src/main/scala/scodec/codecs/package.scala [hunk 3]
```diff
@@ -1017,6 +1062,51 @@ package object codecs {
...
+   *
+   * When decoding, `deMux` is called repeatedly to obtain the next bits (to decode using `valueCodec`) and the
+   * remaining bits (input to `deMux` on next iteration) until a decoding error is encountered or no more bits remain.
+   * The final return value is a list of all decoded element values.
+   *
+   * @param mux element multiplexer
+   * @param deMux element de-multiplexer (should return the next bits to decode and the remaining bits for next iteration)
+   * @param valueCodec element codec (used to decode next bits)
+   * @tparam A element type
+   * @return
```
Comment (mpilquist) [inline src/main/scala/scodec/codecs/package.scala]: Needs `@group combinators` tag

File: src/main/scala/scodec/codecs/package.scala [hunk 4]
```diff
@@ -1017,6 +1062,51 @@ package object codecs {
...
+   * Example:
+   * {{{
+   * val codec = listDelimited(BitVector(' '), ascii)
+   * codec.decode(ascii.encode("i am delimited").require).require.value // List("i", "am", "delimited")
+   * }}}
+   *
+   * @param delimiter the bits used to separate element bit values
+   * @param valueCodec element codec (used to decode next bits)
+   * @tparam A element type
+   * @return
```
Comment (mpilquist) [inline src/main/scala/scodec/codecs/package.scala]: Needs `@group combinators` tag

File: src/main/scala/scodec/codecs/MultiplexedCodec.scala [hunk 5]
```diff
@@ -0,0 +1,100 @@
+package scodec.codecs
+
+import scodec._
+import scodec.bits.BitVector
+
+import scala.language.higherKinds
+
+sealed trait MultiplexedCodec {
```
Comment (mpilquist) [inline src/main/scala/scodec/codecs/MultiplexedCodec.scala]: Add brief comment since this is in public API.

**Comment by mpilquist** [comment]
:+1: Looks good -- just a few documentation comments.

**Comment by ajaychandran** [comment]
@mpilquist Incidentally, `ListCodec` and `VectorCodec` can be removed and the corresponding methods changed to use `MultiplexedCodec`.

**Comment by mpilquist** [comment]
Yeah, though I'm hesitant to do so. The vector/list codecs are used so much that I'd want to ensure there's no performance degradation in using the more general implementation.

**Comment by ajaychandran** [comment]
@mpilquist Hi Michael, I had deleted my suggestion because I realized that the current implementation of  `MultiplexedCodec` cannot be used as a direct replacement (decoding would fail as `decodeSeq` applies the codec for the element repeatedly).

This however brings up an interesting question. Should the inputs (after the first invocation) to repeated `deMux` operations be `remainder bits from decode ++ remainder bits from last deMux`?

This implementation could be used as a replacement where `deMux` would simply return `(bits, BitVector.empty)`. Additionally, the remainder from the last decode can be returned as the remainder for the overall decode.

I am thinking I should push this change anyway as this seems to be the correct (generalized) implementation. Please let me know your thoughts.

**Comment by ajaychandran** [comment]
@mpilquist Performance is adversely affected when using the generic implementation for `decodeSeq` (so we should not use it as a replacement).

MultiplexedCodec.scala
```Scala
  final def decode[F[_], A](dec: Decoder[A], deMux: BitVector => (BitVector, BitVector))(buffer: BitVector)(implicit cbf: collection.generic.CanBuildFrom[F[A], A, F[A]]): Attempt[DecodeResult[F[A]]] = {
    val builder = cbf()
    var temp = deMux(buffer)
    var count = 0
    var error: Option[Err] = None
    while (temp._1.nonEmpty) {
      dec.decode(temp._1) match {
        case Attempt.Successful(DecodeResult(value, remainder)) =>
          builder += value
          count += 1
          temp = deMux(remainder ++ temp._2)
        case Attempt.Failure(err) =>
          error = Some(err.pushContext(count.toString))
          temp = (BitVector.empty, BitVector.empty)
      }
    }
    Attempt.fromErrOption(error, DecodeResult(builder.result(), temp._2))
  }
```

MultiplexedCodecTest.scala
```Scala
      val codec = vectorMultiplexed(
        _ ++ delimiter ++ _,
        bits => (bits, BitVector.empty),
        int32)
```

Roundtrip averages:
  10 - 47604 nanoseconds
  100 - 260616 nanoseconds
  1000 - 17533658 nanoseconds
  10000 - 1991149166 nanoseconds

Notably, the test fails with `StackOverflowError` for a sample size of 100000.
java.lang.StackOverflowError
	at scodec.bits.BitVector$Append.drop(BitVector.scala:1805)
	at scodec.bits.BitVector$Append.drop(BitVector.scala:1805)

I am guessing this is because bits are not being compacted. Should the implementation compact bits at certain (a parameter with a default value?) intervals?

**Comment by ajaychandran** [comment]
@mpilquist Apologies for the multiple posts.

The error was due to non compaction. And compaction (in `deMux`) improved the performance dramatically (though not as good as `decodeSeq`).

MultiplexedCodecTest.scala
```Scala
      val codec = vectorMultiplexed(
        _ ++ delimiter ++ _,
        bits => (bits.compact, BitVector.empty),
        int32)
```

Roundtrip averages:
  10 - 51959 nanoseconds
  100 - 154390 nanoseconds
  1000 - 1500687 nanoseconds
  10000 - 15246380 nanoseconds
  100000 - 170877194 nanoseconds

VectorCodecTest.scala
Roundtrip averages:
  10 - 52567 nanoseconds
  100 - 61703 nanoseconds
  1000 - 483228 nanoseconds
  10000 - 5871514 nanoseconds
  100000 - 56173043 nanoseconds

So the user can decide on whether to compact or not in `deMux`.

Now the only outstanding question is the construction of input bits for `deMux`.

**Comment by mpilquist** [comment]
@ajaychandran Yeah, I agree with your suggestion on the remainder. Let me know when that's in and I'll merge to 1.8.0. I'm planning on releasing 1.8.0 as soon as there's a ScalaTest 3.0.0-Mx build for Scala 2.10.

**Comment by ajaychandran** [comment]
@mpilquist Done. Eagerly awaiting the next release!

**Comment by mpilquist** [comment]
Awesome, thanks!

**Pull request merged**